test/passes/instrument-allocations.txt

```diff
@@ -0,0 +1,82 @@
+(module
+ (type $0 (func (param i32 i32) (result i32)))
+ (type $1 (func (param i32)))
+ (type $2 (func (param i32 i32 i32)))
+ (type $3 (func (param i32) (result i32)))
+ (type $4 (func (result i32)))
+ (type $5 (func (param i32 i32)))
+ (import "env" "malloc" (func $malloc (param i32) (result i32)))
+ (import "env" "realloc" (func $realloc (param i32 i32) (result i32)))
+ (import "env" "calloc" (func $calloc (param i32 i32) (result i32)))
+ (import "env" "free" (func $free (param i32)))
+ (import "env" "trace_calloc" (func $trace_calloc (param i32 i32 i32)))
+ (import "env" "trace_free" (func $trace_free (param i32)))
+ (import "env" "trace_malloc" (func $trace_malloc (param i32 i32)))
+ (import "env" "trace_realloc" (func $trace_realloc (param i32 i32 i32)))
+ (func $test_malloc (result i32)
+  (local $0 i32)
+  (local $1 i32)
+  (call $trace_malloc
+   (local.get $0)
+   (local.tee $1
+    (call $malloc
+     (local.tee $0
+      (i32.const 2)
+     )
+    )
+   )
+  )
+  (local.get $1)
+ )
+ (func $test_realloc (param $0 i32) (param $1 i32) (result i32)
+  (local $2 i32)
+  (local $3 i32)
+  (local $4 i32)
+  (call $trace_realloc
+   (local.get $2)
+   (local.get $3)
+   (local.tee $4
+    (call $realloc
+     (local.tee $2
+      (i32.const 4)
+     )
+     (local.tee $3
+      (i32.const 8)
+     )
+    )
+   )
+  )
+  (local.get $4)
+ )
+ (func $test_calloc (param $0 i32) (param $1 i32) (result i32)
+  (local $2 i32)
+  (local $3 i32)
+  (local $4 i32)
+  (call $trace_calloc
+   (local.get $2)
+   (local.get $3)
+   (local.tee $4
+    (call $calloc
+     (local.tee $2
+      (i32.const 16)
+     )
+     (local.tee $3
+      (i32.const 32)
+     )
+    )
+   )
+  )
+  (local.get $4)
+ )
+ (func $test_free (param $0 i32)
+  (local $1 i32)
+  (call $free
+   (local.tee $1
+    (i32.const 64)
+   )
+  )
+  (call $trace_free
+   (local.get $1)
+  )
+ )
+)
```

Comment:
Another option rather than instrumenting around each function call is to add new wrapper functions. That would be better for code size (existing calls remain the same size, they just point to the wrappers, and the wrappers are just a fixed increase in size). The inliner can inline them if the user then optimizes (and the inliner thinks it is worth it), which could make it just as fast.

But if size/speed are not important here then the current approach may be fine.

Reply:
I personally only intend to use it for debugging and not in the production code (at at least so far) so neither speed nor size is critical. I think the pass can be updated once there are different usecases.